---
title: "panvaR"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{panvaR}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```


# The different ways of using PanvaR

## When you have tag SNP/s handy

`PanvaR` can be run in multiple ways. These different ways of running `panvaR` come down to the type of input you have. The possible inputs to `panvaR` can be divided into optional and required as follows:-

1. Annotated-VCF - A `SnpEff` annotated VCF file, crucial for the `IMPACT` information that `SnpEff` adds to the VCF data.
2. Tag SNP - The chromosome and the BP of the tag SNP. 
  - (Vocabulary) Tag SNP:- What `panvaR` call a tag SNP is the BP around which `panvaR`'s scientific logic will be applied. Usually just the SNP with the highest score from the GWAS.
3. GWAS results - In tabular format - the results of the GWAS for the data that you are working with. This supplies the assocation data for each SNP.

In this case, you can run `panvaR` as
```r
panvar_gwas_tagsnp_snpeff(
  gwas_table_path = "<path/to/gwas/table>",
  vcf_file_path = "<path/to/vcf/file.gz>",
  chrom = "Chr_05",
  bp = 6857045,
  r2_threshold = 0.8
)
```

Details on the inputs:-

### Gwas table

You are free to supply your gwas table but it needs to have at least three fields of data (spelling and case is important):-

1. `BP` - The base-pair loci of the data.
2. `CHROM` - The chromosomes.
3. `Pvalues` - The -log10 of the pvalues for each loci per GWAS.

Example of a table that can be supplied to the `gwas_table_path` parameter.

| Chroms | BP   | Pvalues |
| ------ | ---- | ------- |
| Chr01  | 515  | 15.487  |
| Chr_01 | 2552 | 14.87   |
| chr01  | 4657 | 12.87   |
| 01     | 7889 | 26.72   |

For `panvaR` to run smoothly the above fields should be named exactly as presented - this makes the manipulation of tabular data easier and faster.

### Annotated VCF file

This should be an annotated VCF file from `SpnEff` - if you have a VCF file that is not annotated please use `SnpEff` to annotate it.

### Tag_snp

The tag snp should be supplied with the params `chrom` and `bp`. For example:-

```r
chrom = "Chr_05"
bp = 6857045
```
## When you don't have a tag SNP/s handy

But if you don't have the `tag SNP` you can use the `panvar_gwas` functionality built into `panvaR`. This is a fairly straight forward `GWAS` pipeline that depends on:-

```
library(bigsnpr)
library(modelr)
```

The default behaviour is to use 5 PCs and produce GWAS results using `big_randomSVD`, `big_random_pca`, `big_univLinReg` and `snp_gc` for most of its functionality.

To use `panvaR`'s GWAS module use you can use the following syntax

```
panvar_gwas(
  genotype_data,
  phentotype_path,
  pc_min = 5,
  pc_max = 5, 
  maf = 0.05, 
  missing_rate = 0.1
)
```

If you are looking for details into how the `GWAS` pipeline has been implemented - please read into `R/panvar_gwas.R`. 